odoo/addons/test_runtime/fields.py

```diff
@@ -0,0 +1,41 @@
+from odoo import fields, models
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+class SelectionStr(str):
+    def init_vals(self, selection, field_name):
+        self.__selection = selection
+        self.__field_name = field_name
+
+    def __eq__(self, other):
+        if isinstance(other, str) and other not in self.__selection:
+            _logger.warning(
+                "Wrong check for selection field on %s: %s\n"
+                "Valid values are: %s" % (
+                    self.__field_name,
+                    other,
+                    ', '.join(self.__selection),
+                )
+            )
+        return super().__eq__(other)
+
+    def __hash__(self):
+        return hash(str(self))
+
+class IrModel(models.Model):
+    _inherit = 'ir.model'
+
+    def __init__(self, *args, **kwargs):
+        # Monkey path the convert_to_cache of Selection fields to detect wrong
+        # comparisons. (i.e. deprecated values, typos, wrong field,...)
+        def convert_to_cache(self, value, record, validate=True):
+            if value and IrModel in record.env['ir.model']._model_classes:
+                value = SelectionStr(value)
+                value.init_vals(self.get_values(record.env), '%s.%s' % (record._name, self.name))
+            return convert_to_cache.super(self, value, record, validate)
+        convert_to_cache.super = fields.Selection.convert_to_cache
+        fields.Selection.convert_to_cache = convert_to_cache
```

Comment:
What about monkey-patching the conversion when `config['test-enable']` is set?
This trick requires the module to be installed, and the patch is only effective *after* installation of the module.
Many modules (like `base`) will not be tested at all.